### ISSUE TITLE
Feature: Manage (Add and Remove) NeoStat devices from Home-Assistant

### ISSUE THREAD

**Issue by MindrustUK** [MindrustUK/Heatmiser-for-home-assistant#205]
As mentioned: https://github.com/MindrustUK/Heatmiser-for-home-assistant/issues/60#issuecomment-753622665

It would be nice to be able to manage adding and removing devices directly from Home Assistant.

**Comment by PhillyGilly** [comment]
Supported. Will this include the "phantom repeaters" mention in https://github.com/MindrustUK/Heatmiser-for-home-assistant/issues/196#issuecomment-2448629452

**Comment by ocrease** [comment]
@MindrustUK @PhillyGilly what are we talking about here exactly?

1. Adding and removing devices from Home Assistant so they can/can't be controlled
2. Or actually pairing new devices to the hub so you don't have to use the heatmiser app?

If its the second point, I would argue that this is not something that belongs in this integration or in HA at all. Physically setting up your devices is something that should be done following the heatmiser instructions. Also its something you do once per device and never again? If I compare the experience to say one of my Shelly devices, I can't do the initial device setup from Home Assistant. Once it's set up, I can then add it to home assistant (either manually or via auto discovery) so I can control it from HA. If I want to change advanced settings on the device itself I have to go to it's configuration page. Another example is a new zigbee device. In order to pair it to my coordinator I have to use zigbee2mqtt itself. Once it's paired, the device is added to HA. Do you have an example of an integration where you can do the initial setup of a device via HA?

If it's point 1, then I agree. It should be possible to add a new device to HA once it's been added to the hub. For example if you add a new plug, repeater, temperature or contact sensor to your heatmiser ecosystem, then it should be possible to easily add it to HA. I think at the moment you would have to delete the integration entry and re-add it? Equally it should be possible to remove a device from HA and it should remember that and it shouldn't be added back in automatically. Not sure how this would be handled, but I assume via the config flow.

Regarding repeaters (and possibly all devices), we could consider adding a button to unpair the device from the hub, but if you then wanted to add it to the hub again you would need to go through the heatmiser pairing instructions. (Adding the requirement to unpair can be added to the repeaters ticket)

Just my 2 cents.

**Comment by PhillyGilly** [comment]
I agree 100% that pairing devices to the hub should be a Heatmiser App unique task.
What I was thinking was:

1.  that HA users should be able to add/remove devices from the HA regime, my logic being that there is quite a lot of traffic going on for devices that don't change very often, and some people may only want devices that are used in automations. 
2. that the repeater delete snippet of code should sit in here because not everybody wants to run a bit of python on the side.

I hope that's helpful.

**Comment by MindrustUK** [comment]
@ocrease I think the original intention (as I understood it) by @stikonas is probably to have as much of a FOSS solution as possible, avoiding Heatmiser first party apps where possible. I maybe wrong in this assumption so I'll leave it for @stikonas to clarify. This would Parallel other integrations such as Philip's Hue which can be completely deployed in Home Assistant with no additional software required.

My only horse in this race is on principal, in that I don't like being restricted to the manufacturers application. I'd like to support open solutions wherever possible.

**Comment by stikonas** [comment]
> @ocrease I think the original intention (as I understood it) by @stikonas is probably to have as much of a FOSS solution as possible, avoiding Heatmiser first party apps where possible. I maybe wrong in this assumption so I'll leave it for @stikonas to clarify. This would Parallel other integrations such as Philip's Hue which can be completely deployed in Home Assistant with no additional software required.
> 
> My only horse in this race is on principal, in that I don't like being restricted to the manufacturers application. I'd like to support open solutions wherever possible.

Exactly, I thought that it would be good to have an open solution that does not need proprietary software. Not sure if it's still possible with current Hubs. I keep my Neohub firewalled from internet and I can still pair my thermostats to the hub with no proprietary apps.

>  If I compare the experience to say one of my Shelly devices, I can't do the initial device setup from Home Assistant. 

For shellys, you can flash Tasmota on them. Which is probably the main reason why nobody bothers to reverse engineer Shelly pairing protocol. Why waste time when you can just flash an open firmware.

**Comment by rtb1965** [comment]
I have just added a neostat-hw V2 and managed to get it paired with the neohub v2 but I am struggling to get it to appear in Home Assistant. The hub and 5 neostats V2 for each room with underfloor heating are all working fine in HA. What do I need to do o integrate the new device? Many thanks

**Comment by ocrease** [comment]
![Image](https://github.com/user-attachments/assets/275e0b4b-81e7-47e0-8988-3b45c1d6cba3)

Try reload, or restart home assistant. If that fails, share the version that you are using and upload the diagnostics output (also visible in the screenshot above)

**Comment by rtb1965** [comment]
Perfect - it worked! Thanks so much

**Comment by stikonas** [comment]
@ocrease now added support to add NeoHub itself from Home assistant (using Connect button).

I don't think we have an option to add/remove Neostat devices in HA, though that is possible in neohubapi:

https://gitlab.com/neohubapi/neohubapi/-/blob/master/neohubapi/neohub.py?ref_type=heads#L460

**Comment by ocrease** [comment]
@stikonas removing devices is easy to add, we already support it for repeaters. Adding devices is a bit harder because you have to specify the name of the device when you issue the permit join command. However, I think it could be done via the options flow. I don't have time to mess with my heating system at the moment, but I have a POC on my [fork](https://github.com/ocrease/Heatmiser-for-home-assistant/tree/Pairing) that you are welcome to try out.

It would need cleaning up, but it issues the command to put the hub into pairing mode. You can access it from the configure option of the hub:

![Image](https://github.com/user-attachments/assets/d9771c52-d245-4aec-99af-88fd251f4715)

Once you've done the steps on the thermostat itself, you will need to either restart HA or reload the integration (see my comment from a couple of weeks ago). This of course would need to be cleaned up so it happens automatically.

The Remove button is available on the devices themselves, but it is hidden by default.

![Image](https://github.com/user-attachments/assets/e9cc20c4-db8c-4d8b-be1f-0727dbc658b6)

**Comment by stikonas** [comment]
Thanks, I'll see though I might not have time until late March either...

**Comment by ocrease** [comment]
Release v3.3.0-beta.5 has this new functionality. See the docs for how to use it: https://mindrustuk.github.io/Heatmiser-for-home-assistant/advanced.html

@stikonas I think you were asking for this functionality? Would you be able to try it out? 

@PhillyGilly I think you have some repeaters? If you could try the remove and add flows for repeaters that would be great. I don't have anything other than NeoStat V2s to test it.

**Comment by stikonas** [comment]
I'm struggling to see where removal of devices is implemented in the GUI (I can see where to add them though)

**Comment by ocrease** [comment]
@stikonas as per an earlier comment, it is a button in the diagnostics section of the device page. It is not enabled by default. 

I'll take your feedback as a good thing. I don't want people to use the button by mistake!

**Comment by stikonas** [comment]
Oh yes, should have read earlier posts.

Well, I removed one of my neostats and readded, so far seems to work fine.

**Comment by PhillyGilly** [comment]
I can test the repeaters functionality later this week.

**Comment by ocrease** [comment]
@PhillyGilly that would be great thanks. Note I just release a v3.3.0-beta.6 which has some improvements to the device removal flow

**Comment by stikonas** [comment]
> [@PhillyGilly](https://github.com/PhillyGilly) that would be great thanks. Note I just release a v3.3.0-beta.6 which has some improvements to the device removal flow

Hmm, HACS is refusing to install beta6.

**Comment by ocrease** [comment]
Its also refusing to install beta 4 and 5, so not sure what is going on.

**Comment by ocrease** [comment]
@stikonas I think it's working now. I was able to install beta 6. 

There was a github outage that has just been resolved. I also tried using hacs to install pre release versions of other integrations and got the same error.

**Comment by ocrease** [comment]
@stikonas @PhillyGilly I have released yet another improvement in beta 7 (there is also a beta 8 which you are welcome to try but it is not related to this)

I got rid of the remove buttons. Instead there is a "Remove device" option in the page that lists all the integration devices:

<img width="1108" height="475" alt="Image" src="https://github.com/user-attachments/assets/f5871b3b-daa7-44c0-831c-1ccfe62a5a31" />

Alternatively from the device page itself:

<img width="653" height="192" alt="Image" src="https://github.com/user-attachments/assets/8880432d-202e-41d0-a208-44456c9d8179" />

Another improvement is that if a device is added to the hub via some other method (eg the Heatmiser App), it will automatically be added to the integration

**Comment by ocrease** [comment]
And beta 9... This addresses the case where a device is removed while HA is down

**Comment by PhillyGilly** [comment]
I've got Beta9 running and by coincidence I had to replace a v2 neostat this week (faulty psu which I replaced from Heatmiser for £12.29 delivered). Anycase here is the app without a Hall stat:

<img width="438" height="380" alt="Image" src="https://github.com/user-attachments/assets/863c819b-697a-4395-a45c-c70e15e7851e" />

Adds OK

<img width="687" height="266" alt="Image" src="https://github.com/user-attachments/assets/9a41c14c-f328-4961-b2b4-9885fac6676a" />

<img width="522" height="310" alt="Image" src="https://github.com/user-attachments/assets/8b791fcb-9001-4664-aea5-8c1171c9c33f" />

<img width="374" height="395" alt="Image" src="https://github.com/user-attachments/assets/97b3057c-faa8-4bde-aa35-f6d5085cfcd6" />

**Comment by PhillyGilly** [comment]
Looking at my two repeaters:
<img width="424" height="155" alt="Image" src="https://github.com/user-attachments/assets/2096400c-a7e2-4081-b812-36fbd8fa3fd6" />

<img width="924" height="243" alt="Image" src="https://github.com/user-attachments/assets/86b1d227-ff0d-45af-8c76-b2734fe6ab03" />

I've deleted it by remove device and now I'm re-adding it
<img width="399" height="189" alt="Image" src="https://github.com/user-attachments/assets/9c0e0806-2bbc-4755-ba86-878d4633647e" />
All good. Thank you @ocrease 

Now I have got to rename the device and entity to repeaternode7905 to save changing my dashboards and automations.

**Comment by stikonas** [comment]
So are we all done here then?

**Comment by PhillyGilly** [comment]
Yes. And that's a big thanks from me, as the repeaters were a bit of a pain not being manageable in the app. Now if we had a Heatmiser Zigbee to Wifi bridge .............

**Comment by ocrease** [comment]
@stikonas @PhillyGilly thanks for trying out this version. I'll close this ticket, but note that I'm working on a few other things (unrelated to this feature) so there won't be a proper release immediately.
